java/arcs/core/testutil/Fuzzing.kt

```diff
@@ -0,0 +1,344 @@
+package arcs.core.testutil
+
+import java.time.LocalDateTime
+import kotlin.random.Random
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.runBlocking
+
+/**
+ * This file contains utility classes and methods that establish fuzz-testing infrastructure.
+ *
+ * To write fuzz tests for a unit or assembly:
+ * (1) think of the invariants that make sense for the thing that you want to test.
+ *
+ * For example, if you've built a bank account class, then an invariant might be that the balance
+ * can never be less than 0. (See below for more information on invariants).
+ *
+ * (2) express this invariant as a function, ensuring all non-required structure is externalized.
+ *
+ * An invariant is a statement that's expected to be true, regardless of some context. The idea
+ * here is to make sure that the context (as much as possible) is passed into the invariant
+ * function, rather than choosing a specific context to test.
+ *
+ * For our bank account class, we want to run a sequence of transactions against an account and
+ * then test that the balance isn't negative. The sequence of transactions and the initial balance
+ * are both structure that is incidental to the invariant, so we will make these function
+ * parameters:
+ *
+ * fun invariant_balanceCantGoNegative(initialBalance: Int, transactions: List<Transaction>) {
+ *   val account = Account(initialBalance)
+ *   val transactionsApplied = transactions.map { account.apply(it) }
+ *   assertThat(account.balance).isLargerThan(0)
+ * }
+ *
+ * Generally, the invariant will separate out into three pieces: firstly there may be some
+ * preconditions that need to be checked (for example, we can't guarantee that the balance
+ * will be larger than 0 if the initialBalance is not). Then there will be the body of the
+ * invariant (applying the arguments, running the test, etc.); and finally there will be a
+ * set of postconditions which need to be asserted.
+ *
+ * Splitting the above function up gives us:
+ *
+ * fun invariant_balanceCantGoNegative(initialBalance: Int, transactions: List<Transaction>) {
+ *   // PRECONDITION
+ *   assertThat(initialBalance).isLargerThan(0)
```

Comment:
It's the caller's responsibility. We want it to be an assertion failure rather than an early return because in some cases only a fraction of possible inputs actually meet a particular condition; if we just early returned here then the majority of runs of that test would just be skipping the test.